### ISSUE TITLE
Can this model be used for duplicate detection from both image and text?

### ISSUE THREAD

**Issue by smith-co** [facebookresearch/multimodal#114]
### 🚀 The feature, motivation and pitch

A model for near duplicate detection from both image and text.

Given two pairs of input composed of image and text, determine whether they are semantically duplicate or not.

```
inputA = (imageA, textA)
inputB = (imageB, textB)
```

Determine whether `inputA` and `inputB` are near duplicate or not?

**Comment by RdoubleA** [comment]
Thanks for posting. I am not entirely sure what you mean by semantically duplicate. Do you mean semantic similarity, as in a high cosine similarity between the embeddings of two (image, text) pairs? If so, the CLIP model and Contrastive Loss can help you achieve something close. Do you have a reference of what you had in mind?

**Comment by smith-co** [comment]
@RdoubleA yeah, I am referring whether two (image, text) pairs are semantically similar (near duplicate) or not.

**Comment by RdoubleA** [comment]
I would suggest taking a look at [CLIPArchitecture](https://github.com/facebookresearch/multimodal/blob/main/torchmultimodal/architectures/clip.py) to generate embeddings for your image-text pairs and concatenate them for each input. Then you could use [ContrastiveLossWithTemperature](https://github.com/facebookresearch/multimodal/blob/00380984dc1b9607dfb7ebf34946e5ea0f5bfdf8/torchmultimodal/modules/losses/contrastive_loss_with_temperature.py#L140) and treat the image and text embeddings arguments as inputA embeddings and inputB embeddings. For your use case you should be able to use these existing components from the library.

**Comment by smith-co** [comment]
@RdoubleA thanks for your great insight. I have to initialize the `CLIPArchitecture` as follows:

```
        encoder_a = torch.nn.Linear(5, 3)
        encoder_b = torch.nn.Linear(4, 3)
        clip = CLIPArchitecture(encoder_a, encoder_b)
```

What should I use as the encoders? For text I think I could use SentenceBert, and for image may be I could get embedding from pre-trained ResNet50? For my case, (image, text) pair is very different though.

**Comment by smith-co** [comment]
@RdoubleA also confused why you mentioned that embeddings needs to be concatenated?

**Comment by RdoubleA** [comment]
> @RdoubleA thanks for your great insight. I have to initialize the `CLIPArchitecture` as follows:
> 
> ```
>         encoder_a = torch.nn.Linear(5, 3)
>         encoder_b = torch.nn.Linear(4, 3)
>         clip = CLIPArchitecture(encoder_a, encoder_b)
> ```
> 
> What should I use as the encoders? For text I think I could use SentenceBert, and for image may be I could get embedding from pre-trained ResNet50? For my case, (image, text) pair is very different though.

that seems reasonable to me. I mentioned concatenate because you wanted to compare the embeddings of two pairs. so if you have your image and text encoders, you can use one CLIP to get one pair and another CLIP to get another pair. Then you concatenate image and text within each pair before you use contrastive loss since ContrastiveLossWithTemperature takes two sets of encodings and not four. Or you can compute contrastive loss separately between image embeddings and text embeddings. depending on how you wanted to measure similarity.

**Comment by smith-co** [comment]
Thanks for your response @RdoubleA. I have not really understood how to use [contrastive_loss_with_temperature.py](https://github.com/facebookresearch/multimodal/blob/00380984dc1b9607dfb7ebf34946e5ea0f5bfdf8/torchmultimodal/modules/losses/contrastive_loss_with_temperature.py#L140).

Do I have to implement a siamese network with a contrastive loss function - that's what you meant?

**Comment by RdoubleA** [comment]
Yeah, something similar to that, if you comparing your (image, text) pairs from the same encoders.

ContrastiveLossWithTemperature just takes two embeddings and calculates contrastive loss. It is labeled with image and text but you should be able to do image and image, or text and text, or image+text and image+text if you wanted to. In fact, it might be better on our end to name the parameters to something more general, like in CLIPArchitecture.

**Comment by RdoubleA** [comment]
@smith-co if there's no other questions, I will close this issue.
